doc/api/esm.md

```diff
@@ -183,25 +183,49 @@ versions of Node.js, but its capabilities are limited: it only defines the main
-points are defined in both `"main"` and `"exports"`, the latter takes precedence
-in versions of Node.js that support `"exports"`. [Conditional Exports][] can
-also be used within `"exports"` to define different package entry points per
-environment, including whether the package is referenced via `require` or via
-`import`.
+main entry point can be defined while also encapsulating the package,
+**preventing any other entry points besides those defined in `"exports"`**.
+This encapsulation allows module authors to define a public interface for
+their package 
 
 If both `"exports"` and `"main"` are defined, the `"exports"` field takes
-precedence over `"main"`.
+precedence over `"main"`. It is important to note that `"exports"` are not
+specific to ES modules or CommonJS; `"main"` will be overridden by `"exports"`
+if it exists. As such `"main"` cannot be used as a fallback for CommonJS
+but it can be used as a fallback for legacy versions of Node.js that do not
+support the exports field.
+
+[Conditional Exports][] can be used within `"exports"` to define different
+package entry points per environment, including whether the package is
+referenced via `require` or via `import`. For more information about supporting
+both Common.js and ES Modules in a single package please consult
+[the dual CommonJS/ES module packages section][]. 
+
+**Warning**: Introducing the `"exports"` field prevents consumers of a package
+from using any entry points that are not defined, including the `package.json`.
+This will likely be a breaking change. To make the introduction of `"exports"`
+non-breaking, ensure that every previously supported entry point is exported.
+It is best to explicitly specify entry points so that the package's public API
+is well-defined. For example, a project that previous exported `main`, `lib`,
+`feature`, and the `package.json` could use the following `package.exports`:
 
-Both `"main"` and `"exports"` entry points are not specific to ES modules or
-CommonJS; `"main"` will be overridden by `"exports"` in a `require` so it is
-not a CommonJS fallback.
+```json
+{
+  "exports": {
+    ".": "./lib/index.js",
+    "./lib": "./lib/index.js",
+    "./lib/index.js": "./lib/index.js",
+    "./feature": "./feature/index.js",
+    "./feature/index.js": "./feature/index.js",
+    "./package.json": "./package.json"
+  }
+}
+```
 
-This is important with regard to `require`, since `require` of ES module files
-throws an error in all versions of Node.js. To create a package that works both
-in modern Node.js via `import` and `require` and also legacy Node.js versions,
-see [the dual CommonJS/ES module packages section][].
+As a last resort, package encapsulation can be disabled entirely by creating an
+export for the root of the package `"./": "./"`. This will expose every file in
+the package at the cost of disabling encapsulation and the tooling and optimization
+benefits that come with it.
```

Comment:
Since these tooling and optimization benefits don't yet exist, it feels a bit premature to point to them as a reason to use `"exports"`. This paragraph comes off a bit as "we at Node want you to use `"exports"`, and to use encapsulation, but we aren't going to really explain why."

There is a straightforward use case for `"./": "./"` (versus not adding `"exports"` at all): a package author who doesn't want encapsulation but who wants to use conditional exports to define separate entry points for CommonJS and ESM. One could argue that `"./": "./"` is the safer choice as it's for-sure not a breaking change, versus spelling out every export and possibly missing one.

Reply:
I added this phrasing at the request of @guybedford would you like to suggest an alternative?

Reply:
I've removed the copy about tooling

Reply:
> Since these tooling and optimization benefits don't yet exist

All my own tooling very much relies on and uses exports for automatic optimization, including the next version of jspm. And I always opt-out of all of these optimizations when `./`: `./` is set.

I would have preferred we included this line as when giving advice it is always helpful to explain the reason for the advice.

That said I understand there has been a silly amount of back and fourth there so I won't press this.